### PR TITLE
feat(unlock-app) - handling non-convertible currencies for credit card payments

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description

**Checkout with custom USD price**

Screenshot 2023-05-05 at 20 02 40

**Checkout with custom USD price**

Screenshot 2023-05-05 at 20 02 33

**Set credit card price for `exotic` currency**

Screenshot 2023-05-04 at 20 03 09

**Set credit card price base currency**

Screenshot 2023-05-04 at 20 32 28

**Validation for value**

Screenshot 2023-05-04 at 19 50 09

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 1]
```diff
@@ -40,19 +55,164 @@ interface DisconnectStripeProps {
...
+                setValue(
+                  'creditCardPrice',
+                  enabled ? lockSettings?.creditCardPrice : null,
+                  {
+                    shouldValidate: true,
+                  }
+                )
+              }}
+            />
+          )}
```
Comment (kalidiagne) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:167]: show toggle to set custom price only when a conversion is present

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 2]
```diff
@@ -40,19 +55,164 @@ interface DisconnectStripeProps {
...
+          {(useCustomPrice || !hasPriceConversion) && (
+            <>
+              <Input
+                type="numeric"
+                step="any"
+                disabled={disabled}
+                description={
+                  hasPriceConversion
+                    ? 'This value will replace the actual conversion price when set. To use default conversion'
+                    : `This value will be used as default because there is no conversion price for ${symbol}`
```
Comment (kalidiagne) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:177]: @julien51 can we improve copy here?

File: unlock-app/src/hooks/useCreditCard.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  lockAddress,
+  network,
+}: CreditCardProps) => {
+  const config = useConfig()
+
+  return useQuery(['getCreditCardPricing', lockAddress, network], async () => {
+    const keyPricing = await getFiatPricing(config, lockAddress, network)
+    return keyPricing
+  })
+}
```
Comment (searchableguy) [inline unlock-app/src/hooks/useCreditCard.ts:21]: Don't we already have a hook for this? usePricing

Comment (kalidiagne) [reply unlock-app/src/hooks/useCreditCard.ts:21]: I need credit card converted price here, and `usePricing` does not return this value

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 4]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
+          <>
+            {(useCustomPrice || !hasPriceConversion) && (
+              <Input
+                type="numeric"
+                step="any"
+                disabled={disabled}
+                description={
+                  hasPriceConversion
+                    ? 'Price in $USD that will be charged, the actual conversion price will ignored.'
+                    : `Price in $USD that will be charged`
```
Comment (kalidiagne) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:179]:
```suggestion
                  hasPriceConversion
                    ? 'Price in $USD that will be charged, the actual conversion price will ignored.'
                    : `Price in $USD that will be charged`
```

@julien51 can we improve copy here?

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 5]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
+                }
```
Comment (kalidiagne) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:180]: @julien51 can we improve copy here?

File: locksmith/src/utils/keyPricer.ts [hunk 6]
```diff
@@ -56,6 +57,16 @@ export default class KeyPricer {
     if (lock.keyPrice === '0') {
       return 0
     }
+
+    const { creditCardPrice } = await lockSettingOperations.getSettings({
```
Comment (julien51) [inline locksmith/src/utils/keyPricer.ts:61]: Is `keyPriceUSD` the function used when processing actual payments? I want to make sure that if we merge that code, we will actually use the settings' `creditCardPrice` (and not a conversion)

File: locksmith/src/utils/keyPricer.ts [hunk 7]
```diff
@@ -56,6 +57,16 @@ export default class KeyPricer {
...
+      lockAddress,
+      network,
+    })
+
+    if (creditCardPrice) {
+      return parseInt((creditCardPrice * 100).toFixed(0)) // returns the USD cents price
```
Comment (julien51) [inline locksmith/src/utils/keyPricer.ts:67]: CAn you confirm the same format is used by `priceConversion.convertToUSD` ?

Comment (kalidiagne) [reply locksmith/src/utils/keyPricer.ts:67]: Yes, is the same format!

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 8]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
+                      shouldValidate: true,
+                      shouldDirty: true,
+                    })
+                  }
+                }}
+              />
+              {!useCustomPrice && (
+                <span className="text-sm">
+                  When custom price is not set, the conversion price will be
+                  used for credit card payments.
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:166]:
```suggestion
                  If you set a custom price for card payments, we will use that one (+ fees) instead of converting the price of `symbol` to USD.
```

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx (hunk 5 above)
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:181]:
```suggestion
                description="Set a fixed price in USD that is charged for card payments."
```

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 9]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
+                  hasPriceConversion
+                    ? 'Price in $USD that will be charged, the actual conversion price will ignored.'
+                    : `Price in $USD that will be charged`
+                }
+                error={errors?.creditCardPrice?.message}
+                {...register('creditCardPrice', {
+                  required: true,
+                  min: {
+                    value: 0.5,
+                    message:
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx]: Can we actually use `errors?.creditCardPrice?.message` here for clarity so that if the error is something else we know about it?

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 10]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
+              <Button className="w-full" size="small" disabled={saveDisabled}>
+                Apply
+              </Button>
+            </div>
+          </>
+        </form>
+      </div>
+      {needsCustomPrice && (
+        <span className="text-sm font-semibold text-red-600">
+          {`There is no conversion price for ${symbol} your lock. Your price needs to be set manually.`}
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:204]:
```suggestion
          We are not able to convert  {symbol} to USD. Please set a fixed price you want to charge for card payments.
```

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 11]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
   return (
     <div className="flex flex-col gap-4">
-      <span className="text-xs">
-        <SettingCardDetail
-          title="Credit card payment ready"
-          description="Member of this Lock can now pay with credit card or crypto as they wish. "
-        />
-      </span>
+      <SettingCardDetail
+        title="Credit card payment ready"
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:219]:
```suggestion
        title="Card payments enabled"
```

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx [hunk 12]
```diff
@@ -40,19 +54,169 @@ interface DisconnectStripeProps {
...
+        description="Member of this Lock can now pay with credit card or crypto as they wish. "
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx:220]:
```suggestion
        description="Your members can purchase their memberships using a card or their crypto wallets."
```

File: unlock-app/src/hooks/useLockSettings.ts [hunk 13]
```diff
@@ -70,3 +77,25 @@ export function useSaveSlugSetting() {
     }
   })
 }
+
+export function useSaveLockSettings() {
```
Comment (julien51) [inline unlock-app/src/hooks/useLockSettings.ts:81]: Please make sure you replace all the other places where we have the same API call?

File: unlock-app/src/hooks/useLockSettings.ts [hunk 14]
```diff
@@ -70,3 +77,25 @@ export function useSaveSlugSetting() {
...
+  return useMutation(async (settings: SettingsProps) => {
+    const { lockAddress, network } = settings
+
+    return storage.saveLockSetting(network, lockAddress, {
+      ...settings,
+    })
+  })
+}
+
+export function useGetLockSettings({
```
Comment (julien51) [inline unlock-app/src/hooks/useLockSettings.ts:91]: Same here.
Also, let's actually move that to a different PR please.

File: unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx (hunk 9 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/locks/Settings/forms/CreditCardForm.tsx]: error message is already defined in `error` prop and this is the specific message for `min` value

File: unlock-app/src/hooks/useLockSettings.ts (hunk 13 above)
Comment (kalidiagne) [reply unlock-app/src/hooks/useLockSettings.ts:81]: i will move in a different PR

File: unlock-app/src/hooks/useLockSettings.ts (hunk 14 above)
Comment (kalidiagne) [reply unlock-app/src/hooks/useLockSettings.ts:91]: i will move in a different PR

File: locksmith/src/utils/keyPricer.ts (hunk 6 above)
Comment (kalidiagne) [reply locksmith/src/utils/keyPricer.ts:61]: this is related for the FE part, i also added the part to pass the correct value for the `purchase` intent for stripe

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 15]
```diff
@@ -122,17 +122,23 @@ export function PricingData({ pricingData, lock, payment }: PricingDataProps) {
...
-                        .convertToQuoteToken(item.amount.toString())
-                        .toFixed()
-                    ).toLocaleString()} ${symbol}`
-                  : `${formatNumber(item.amount).toLocaleString()} ${symbol}`}
-              </div>
+              {item?.creditCardAmount > 0 ? (
+                <div className="font-bold">
+                  {`$${formatNumber(item.creditCardAmount).toLocaleString()}`}
+                </div>
+              ) : (
```
Comment (kalidiagne) [inline unlock-app/src/components/interface/checkout/main/Confirm.tsx:129]: show credit card pricing when present

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 16]
```diff
@@ -122,17 +122,23 @@ export function PricingData({ pricingData, lock, payment }: PricingDataProps) {
...
+                    ? 'FREE'
+                    : payment?.route
+                    ? `${formatNumber(
+                        payment.route
+                          .convertToQuoteToken(item.amount.toString())
+                          .toFixed()
+                      ).toLocaleString()} ${symbol}`
+                    : `${formatNumber(item.amount).toLocaleString()} ${symbol}`}
+                </div>
+              )}
```
Comment (kalidiagne) [inline unlock-app/src/components/interface/checkout/main/Confirm.tsx:141]: @julien51  

Here we show the credit card price for the single key in $USD, if present 

Screenshot 2023-05-08 at 16 36 50

otherwise, shoe the price by `symbol` price

Screenshot 2023-05-08 at 16 19 58

File: locksmith/src/utils/keyPricer.ts (hunk 6 above)
Comment (julien51) [reply locksmith/src/utils/keyPricer.ts:61]: Actually no this is not related to the front-end but to the backend!
We need to make sure that the *same* function on the backend is used to return the price AND to charge that amount.

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx (hunk 16 above)
Comment (julien51) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:141]: I don't understand why we show it differently?
We should always show the $ prices here since that is what the user is going to pay. We should not show MATIC on that screen at all.

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 17]
```diff
@@ -175,7 +181,7 @@ export function Confirm({
 
   const currencyContractAddress = swap
     ? payment.route.trade.inputAmount.currency?.address
-    : lock?.currencyContractAddress
+    : lock?.currencyContractAddress ?? undefined
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/main/Confirm.tsx:184]: This is very confusing. Any reason why you set it to `undefined`? `lock?.currencyContractAddress` should already be undefined or null if unset

File: unlock-app/src/components/interface/locks/Settings/elements/SettingCard.tsx [hunk 18]
```diff
@@ -11,7 +11,7 @@ interface SettingCardProps {
 
 interface SettingCardDetailProps {
   title: string
-  description: ReactNode
+  description?: ReactNode
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/elements/SettingCard.tsx:14]: Let's remove that change from this PR

**Review by julien51 — COMMENTED** [review]
There is a lot going on here. Can you split in smaller PRs that would be more incremental please?

Maybe the backend changes first thing everything should work the same once they are merged (no lock has a hardcoded price) and then we can look at the front-end changes?

**Review by julien51 — COMMENTED** [review]
And even the front-end changes we can split them:

1. the checkout changes: we can verify that nothing breaks because no lock will have a custom pricing
2. finally the settings changes and we can see that all changes are reflected when a hardcoded price is set?

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx (hunk 16 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:141]: > I don't understand why we show it differently? We should always show the $ prices here since that is what the user is going to pay. We should not show MATIC on that screen at all.

As for now, we always show the `symbol` and in the `credit card section` the price breakdown with the conversion.
I can change it and always keep the `$USD` price when converted or have credit card price

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx (hunk 17 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:184]: Is passing the contract address -> `'null'` as a string, and this is not right

File: locksmith/src/utils/keyPricer.ts (hunk 6 above)
Comment (kalidiagne) [reply locksmith/src/utils/keyPricer.ts:61]: > Actually no this is not related to the front-end but to the backend! We need to make sure that the _same_ function on the backend is used to return the price AND to charge that amount.



> There is a lot going on here. Can you split in smaller PRs that would be more incremental please?
> 
> Maybe the backend changes first thing everything should work the same once they are merged (no lock has a hardcoded price) and then we can look at the front-end changes?

ok, let me split in different PR

**Comment by kalidiagne** [comment]
closed by https://github.com/unlock-protocol/unlock/pull/12183
